lib/bolt/pal.rb

```diff
@@ -70,134 +67,38 @@ def initialize(modulepath, hiera_config, resource_types, max_compiles = Etc.npro
-        # Windows 'fix' for openssl behaving strangely. Prevents very slow operation
-        # of random_bytes later when establishing winrm connections from a Windows host.
-        # See https://github.com/rails/rails/issues/25805 for background.
-        require 'openssl'
-        OpenSSL::Random.random_bytes(1)
-      end
-
-      begin
-        require 'puppet_pal'
-      rescue LoadError
-        raise Bolt::Error.new("Puppet must be installed to execute tasks", "bolt/puppet-missing")
-      end
-
-      require 'bolt/pal/logging'
-      require 'bolt/pal/issues'
-      require 'bolt/pal/yaml_plan/loader'
-      require 'bolt/pal/yaml_plan/transpiler'
-
-      # Now that puppet is loaded we can include puppet mixins in data types
-      Bolt::ResultSet.include_iterable
-    end
-
-    def setup
-      unless @loaded
-        # This is slow so don't do it until we have to
-        Bolt::PAL.load_puppet
-
-        # Make sure we don't create the puppet directories
-        with_puppet_settings { |_| nil }
-        @loaded = true
-      end
-    end
-
-    # Create a top-level alias for TargetSpec and PlanResult so that users don't have to
-    # namespace it with Boltlib, which is just an implementation detail. This
-    # allows them to feel like a built-in type in bolt, rather than
-    # something has been, no pun intended, "bolted on".
-    def alias_types(compiler)
-      compiler.evaluate_string('type TargetSpec = Boltlib::TargetSpec')
-      compiler.evaluate_string('type PlanResult = Boltlib::PlanResult')
-    end
-
-    # Register all resource types defined in $Project/.resource_types as well as
-    # the built in types registered with the runtime_3_init method.
-    def register_resource_types(loaders)
-      static_loader = loaders.static_loader
-      static_loader.runtime_3_init
-      if File.directory?(@resource_types)
-        Dir.children(@resource_types).each do |resource_pp|
-          type_name_from_file = File.basename(resource_pp, '.pp').capitalize
-          typed_name = Puppet::Pops::Loader::TypedName.new(:type, type_name_from_file)
-          resource_type = Puppet::Pops::Types::TypeFactory.resource(type_name_from_file)
-          loaders.static_loader.set_entry(typed_name, resource_type)
-        end
-      end
-    end
-
-    def detect_project_conflict(project, environment)
-      return unless project && project.load_as_module?
-      # The environment modulepath has stripped out non-existent directories,
-      # so we don't need to check for them
-      modules = environment.modulepath.flat_map do |path|
-        Dir.children(path).select { |name| Puppet::Module.is_module_directory?(name, path) }
-      end
-      if modules.include?(project.name)
-        Bolt::Logger.warn_once("project shadows module",
-                               "The project '#{project.name}' shadows an existing module of the same name")
-      end
+      @compiler_service = Bolt::PAL::CompilerService.new(project, @modulepath)
```

Comment:
Is this a chance to refactor the PAL class so the initializer isn't the thing that takes the modulepath, but instead we pass instances of applicator and compiler service?

Something we talked about in slack yesterday with respect to how PE builds on top of Bolt::PAL is that one of the things we need to override in PE::PAL is the Bolt::PAL initializer, and we only need to do that because we need to control the modulepath differently.

In that discussion you mentioned:
> ideally the applicator would probably be an argument to the PAL constructor and then we could avoid this situation entirely

Potentially I feel like that applies here: we can avoid the modulepath/user_modulepath logic if we just have the init method for PAL take instances of CompilerService and Applicator as params.